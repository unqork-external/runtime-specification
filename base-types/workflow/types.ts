export type Step = {
  currentStep: boolean
  name: string
  path: string
  percentage: number
  percentageForGroup: number
  navigationState: Record<string, any>
  stepNumber: number
  stepNumberForGroup: number
  lastStep: boolean
}

export type WorkflowNavigation = {
  disableNavigationButtons: boolean
  displayType: string
  enableVerticalNavDisplay: boolean
  goto: boolean
  next: string
  prev: string
  path: string
  stateControlType: string
  step: Step
  steps: Step[]
}

export enum WorkflowNavigationActionType {
  NEXT = 'next',
  PREV = 'prev',
  GOTO = 'goto',
}
