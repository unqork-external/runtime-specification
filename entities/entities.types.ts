import { deserialize } from '@tsed/json-mapper'
import { CollectionOf, Default, Description, Example, Optional, Required } from '@tsed/schema'

import { EntityTarget } from './entitiesTarget.types'

export const enum EntityType {
  SUBMISSION_DATA = 'submission_data',
  DATA_MODEL = 'data_model',
}

@Description('An Entity is a data structure within the engine to represent a user-defined model.')
export class Entity {
  @Optional()
  @Default(EntityType.SUBMISSION_DATA)
  @Description('What type of entity is this? Used to group related entities')
  type?: EntityType = EntityType.SUBMISSION_DATA

  @Required()
  @Example('moduleId')
  @Description('Entity key is a unique key to target entities')
  key: string

  @Required()
  @Example('12345')
  @Description('The module id the entity is associated with.')
  moduleId: string

  @Required()
  @Example('[{key: firstName, property: firstName }]', `[{key: firstName, property: differentPropName }]`)
  @Description('Targets are a collection of engine value references used get and set engine values')
  @CollectionOf(EntityTarget)
  targets: EntityTarget[]
}

export const createEntity = (obj?: Partial<Entity>) => {
  return deserialize({ ...obj }, { type: Entity })
}
