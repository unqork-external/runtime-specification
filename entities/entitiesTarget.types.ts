import { Description, Example, Optional, Required, CollectionOf } from '@tsed/schema'

import { trimAll } from '../../utilities'

@Description(
  trimAll(`
  An Entity Target is a representation of a key that maps to an engine value and a property that maps to the Entity data.
`),
)
export class EntityTarget {
  @Required()
  @Example('firstName', 'grid', 'checkboxOptionA')
  @Description('The name/key of the target that maps to a engine value')
  keyName: string

  @Required()
  @Example('{{firstName.value}}', 'true', 'unknownValue')
  @Description('Either the explicit value for the entity or an interpolated path to resolve the entity value')
  value: string

  @Optional()
  @Example('[{keyName: "firstName", value: "{{firstName.value}}" }]', `[{keyName: "optionA", value: true" }]`)
  @Description(
    trimAll(`
    Targets are an optional collection of engine value references used get and set engine values. 
    An Entity target might optionally need to define exactly what values it cares about. 
    This is useful when dealing with nested data
  `),
  )
  @CollectionOf(EntityTarget)
  targets?: EntityTarget[]
}
