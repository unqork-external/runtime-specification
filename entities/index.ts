export * from './entities.types'
export * from './entitiesTarget.types'
