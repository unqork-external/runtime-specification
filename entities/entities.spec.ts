import { Entity, createEntity } from './entities.types'
import { type EntityTarget } from './entitiesTarget.types'
import { generateSchemaAndValidate } from '../../utilities'

describe('Entities Spec', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(Entity)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toMatchSnapshot()
  })

  it('should pass validation', () => {
    const entity = createEntity({
      key: 'moduleId',
      moduleId: '12345',
      targets: [
        {
          keyName: 'firstName',
          value: 'firstName.value',
        },
      ],
    })
    const isValid = validate(entity)
    expect(isValid).toBe(true)
    const errors = validate.errors

    expect(errors).toStrictEqual(null)
  })

  it('should pass validation with nested targets targets', () => {
    const entity = createEntity({
      key: 'moduleId',
      moduleId: '12345',
      targets: [
        {
          keyName: 'firstName',
          value: 'firstNameValue',
        },
        {
          keyName: 'grid',
          value: 'gridValue',
          targets: [
            {
              keyName: 'lastName',
              value: 'lastNameValue',
            },
          ],
        },
      ],
    })
    const isValid = validate(entity)
    expect(isValid).toBe(true)
    const errors = validate.errors

    expect(errors).toStrictEqual(null)
  })

  it('should pass validation with empty targets', () => {
    const entity = createEntity({ key: 'moduleId', moduleId: '12345', targets: [] })
    const isValid = validate(entity)
    expect(isValid).toBe(true)
    const errors = validate.errors

    expect(errors).toStrictEqual(null)
  })

  it('should fail validation with empty object', () => {
    const entity = {}
    const isValid = validate(entity)
    expect(isValid).toBe(false)
    const errors = validate.errors

    expect(errors).toStrictEqual([
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'key'",
        params: { missingProperty: 'key' },
        schemaPath: '#/required',
      },
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'moduleId'",
        params: { missingProperty: 'moduleId' },
        schemaPath: '#/required',
      },
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'targets'",
        params: { missingProperty: 'targets' },
        schemaPath: '#/required',
      },
    ])
  })

  it('should fail validation with no targets', () => {
    const entity = createEntity({ key: 'moduleId' })
    const isValid = validate(entity)
    expect(isValid).toBe(false)
    const errors = validate.errors

    expect(errors).toStrictEqual([
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'moduleId'",
        params: { missingProperty: 'moduleId' },
        schemaPath: '#/required',
      },
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'targets'",
        params: { missingProperty: 'targets' },
        schemaPath: '#/required',
      },
    ])
  })

  it('should fail validation with no key', () => {
    const entity = createEntity({ targets: [] })
    const isValid = validate(entity)
    expect(isValid).toBe(false)
    const errors = validate.errors

    expect(errors).toStrictEqual([
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'key'",
        params: { missingProperty: 'key' },
        schemaPath: '#/required',
      },
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'moduleId'",
        params: { missingProperty: 'moduleId' },
        schemaPath: '#/required',
      },
    ])
  })

  it('should fail validation with property missing from target', () => {
    const entity = createEntity({ key: 'moduleId', targets: [{ keyName: 'firstName' }] as EntityTarget[] })
    const isValid = validate(entity)
    expect(isValid).toBe(false)
    const errors = validate.errors

    expect(errors).toStrictEqual([
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'moduleId'",
        params: {
          missingProperty: 'moduleId',
        },
        schemaPath: '#/required',
      },
      {
        instancePath: '/targets/0',
        keyword: 'required',
        message: "must have required property 'value'",
        params: {
          missingProperty: 'value',
        },
        schemaPath: '#/required',
      },
    ])
  })

  it('should fail validation with key missing from target', () => {
    const entity = createEntity({ key: 'moduleId', targets: [{ value: 'firstName' }] as EntityTarget[] })
    const isValid = validate(entity)
    expect(isValid).toBe(false)
    const errors = validate.errors

    expect(errors).toStrictEqual([
      {
        instancePath: '',
        keyword: 'required',
        message: "must have required property 'moduleId'",
        params: {
          missingProperty: 'moduleId',
        },
        schemaPath: '#/required',
      },
      {
        instancePath: '/targets/0',
        keyword: 'required',
        message: "must have required property 'keyName'",
        params: {
          missingProperty: 'keyName',
        },
        schemaPath: '#/required',
      },
    ])
  })
})
