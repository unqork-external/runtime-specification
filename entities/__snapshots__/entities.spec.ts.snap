// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entities Spec should match this JSON Schema 1`] = `
{
  "definitions": {
    "EntityTarget": {
      "description": "An Entity Target is a representation of a key that maps to an engine value and a property that maps to the Entity data.",
      "properties": {
        "keyName": {
          "description": "The name/key of the target that maps to a engine value",
          "examples": [
            "firstName",
            "grid",
            "checkboxOptionA",
          ],
          "minLength": 1,
          "type": "string",
        },
        "targets": {
          "description": "Targets are an optional collection of engine value references used get and set engine values. An Entity target might optionally need to define exactly what values it cares about. This is useful when dealing with nested data",
          "examples": [
            "[{keyName: "firstName", value: "{{firstName.value}}" }]",
            "[{keyName: "optionA", value: true" }]",
          ],
          "items": {
            "$ref": "#/definitions/EntityTarget",
          },
          "type": "array",
        },
        "value": {
          "description": "The explicit path to the value for the entity",
          "examples": [
            "firstName.value",
            "true",
            "unknownValue",
          ],
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "keyName",
        "value",
      ],
      "type": "object",
    },
  },
  "description": "An Entity is a data structure within the engine to represent a user-defined model.",
  "properties": {
    "key": {
      "description": "Entity key is a unique key to target entities",
      "examples": [
        "moduleId",
      ],
      "minLength": 1,
      "type": "string",
    },
    "moduleId": {
      "description": "The module id the entity is associated with.",
      "examples": [
        "12345",
      ],
      "minLength": 1,
      "type": "string",
    },
    "targets": {
      "description": "Targets are a collection of engine value references used get and set engine values",
      "examples": [
        "[{key: firstName, property: firstName }]",
        "[{key: firstName, property: differentPropName }]",
      ],
      "items": {
        "$ref": "#/definitions/EntityTarget",
      },
      "type": "array",
    },
    "type": {
      "default": "submission_data",
      "description": "What type of entity is this? Used to group related entities",
      "type": "string",
    },
  },
  "required": [
    "key",
    "moduleId",
    "targets",
  ],
  "type": "object",
}
`;
